refactor(worker): remove storing template object on the job entity

scopsy opened the pull request:
- Resuce the mongo job stored size object by removing the duplicate template entity

### What changed? Why was the change needed?

<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->

### Screenshots

<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>

scopsy committed:
Add StepTemplateFetcher service and refactor step template retrieval

Introduces the StepTemplateFetcher service to centralize fetching of step templates from the database. Refactors SendMessage use case to use this service instead of relying on in-memory step templates, improving data consistency. Updates dependency injection and module providers accordingly. Also adjusts CreateNotificationJobs to only retain the 'type' property in step templates for jobs.

scopsy committed:
Update send-message.usecase.ts

scopsy committed:
Refactor step ID handling in workflow services

Updated step ID checks in StepTemplateFetcher and SendMessage use case to consistently use _id for step identification. Also limited organizationRepository query in StandardWorker to only fetch the _id field for efficiency.

scopsy committed:
Update step-template-fetcher.service.ts

scopsy committed:
Refactor step template fetching to return step only

Updated StepTemplateFetcher to return only the NotificationStepEntity instead of both the step and its template. Adjusted SendMessage use case to accommodate this change, simplifying the step data structure and removing unnecessary template fetching logic.

scopsy committed:
Update send-message.usecase.ts

scopsy committed:
Update send-message.usecase.ts

scopsy committed:
Update reusable-api-e2e.yml

djabarovgeorge commented on apps/worker/src/app/workflow/services/step-template-fetcher.service.ts:
```diff
@@ -0,0 +1,55 @@
+import { NotificationStepEntity, NotificationTemplateRepository } from '@novu/dal';
+
+@Injectable()
+export class StepTemplateFetcher {
```
personal preference not to use the naming template as it can confuse. 
```suggestion
export class WorkflowStepFetcher {
```

djabarovgeorge commented on apps/worker/src/app/workflow/usecases/send-message/send-message.usecase.ts:
```diff
@@ -88,7 +90,31 @@ export class SendMessage {
+    const stepType = command.step.template?.type;
+
+    let stepTemplateResult;
+    if (command.step.template?.type !== StepTypeEnum.TRIGGER) {
```
```suggestion
    if (stepType !== StepTypeEnum.TRIGGER) {
```

djabarovgeorge commented on apps/worker/src/app/workflow/usecases/send-message/send-message.usecase.ts:
```diff
@@ -88,7 +90,31 @@ export class SendMessage {
+
+      stepTemplateResult = await this.stepTemplateFetcher.fetchStepTemplate({
+        workflowId: command._templateId,
+        stepId: stepId || '',
```
strange did TS yield that it can be undefined after the if condition above?
```suggestion
        stepId,
```

djabarovgeorge commented on apps/worker/src/app/workflow/services/step-template-fetcher.service.ts:
i wonder if it could be a function, because i see it used only in send-message.usecase.ts

djabarovgeorge reviewed (commented):
id suggest making this change more explicit because right now we override existing, and for that we will need to:
1. remove the step SendMessageCommand.
2. remove the step from JobEntity.
3. perhaps add a new param stepType instead of JobEntity.step (flatten the step data).

without that, it is hard to say if there are other places in the code that could be affected by this change.

scopsy committed:
Merge branch 'next' into fix-stop-storing-template-on-step

scopsy committed:
Refactor step handling in job execution flow

Removed StepTemplateFetcher service and updated job execution to pass stepId and stepType instead of the full step object. Adjusted related use cases and command definitions to reflect this change, simplifying step data handling and improving clarity.

scopsy committed:
Update send-message.usecase.ts

scopsy committed:
Update index.ts

scopsy committed:
Refactor job step entity and step handling logic

Updated the JobEntity to use a simplified step object instead of NotificationStepEntity. Adjusted related use cases and repositories to reference the new step structure and improved error handling and variable normalization logic. This refactor streamlines step identification and template handling for jobs.

scopsy committed:
Merge branch 'next' into fix-stop-storing-template-on-step

scopsy committed:
Merge branch 'next' into fix-stop-storing-template-on-step
